npm/generate_katas_content.js

```diff
@@ -60,174 +49,250 @@ function getTitleFromMarkdown(markdown) {
-    join(path, contentFileNames.index),
-    "utf8"
+function generateMacroSection(kataPath, match, globalCodeSources) {
+  const type = match.groups.type;
+  const propertiesJson = match.groups.json;
+  const properties = tryParseJSON(
+    propertiesJson,
+    `Invalid JSON for ${type} macro.`
   );
-  const contentAsHtml = marked.parse(contentAsMarkdown);
-  const title = getTitleFromMarkdown(contentAsMarkdown);
-  return {
-    type: "reading",
-    id: id,
-    title: title,
-    contentAsMarkdown: contentAsMarkdown,
-    contentAsHtml: contentAsHtml,
-  };
-}
-
-function symmetricDifference(setA, setB) {
-  const difference = new Set(setA);
-  for (const elem of setB) {
-    if (difference.has(elem)) {
-      difference.delete(elem);
-    } else {
-      difference.add(elem);
-    }
+  if (type === "example") {
+    return generateExampleSection(kataPath, properties);
+  } else if (type === "exercise") {
+    return generateExerciseSection(kataPath, properties, globalCodeSources);
   }
-  return difference;
+
+  throw new Error(`Unknown macro type ${type}`);
 }
 
-function getItemType(path) {
-  const itemTypeFileSets = {
-    reading: new Set([contentFileNames.index]),
-    example: new Set([contentFileNames.index, contentFileNames.qsharpExample]),
-    exercise: new Set([
-      contentFileNames.index,
-      contentFileNames.qsharpPlaceholder,
-      contentFileNames.qsharpSolution,
-      contentFileNames.qsharpVerify,
-      contentFileNames.textSolution,
-    ]),
+function generateTextSection(markdown) {
+  const html = marked.parse(markdown);
+  return {
+    type: "text",
+    contentAsMarkdown: markdown,
+    contentAsHtml: html,
   };
-
-  const itemFiles = new Set(readdirSync(path));
-  return Object.keys(itemTypeFileSets).find(function (key) {
-    const fileSet = itemTypeFileSets[key];
-    return symmetricDifference(fileSet, itemFiles).size === 0;
-  });
 }
 
-function buildItemContent(path) {
-  const itemId = `${basename(dirname(path))}__${basename(path)}`;
-  const itemType = getItemType(path);
-  if (itemType === "example") {
-    return buildExampleContent(itemId, path);
-  } else if (itemType === "exercise") {
-    return buildExerciseContent(itemId, path);
-  } else if (itemType === "reading") {
-    return buildReadingContent(itemId, path);
+function generateSections(kataPath, markdown, globalCodeSources) {
+  const sections = [];
+  const macroRegex = /@\[(?<type>\w+)\]\(\s*(?<json>\{[^@]+\})\s*\)\s+/g;
+  let latestProcessedIndex = 0;
+  while (latestProcessedIndex < markdown.length) {
+    const match = macroRegex.exec(markdown);
+    if (match !== null) {
+      // If there is something between the last processed index and the start of the match, create a text section for
+      // it.
+      const delta = match.index - latestProcessedIndex;
+      if (delta > 0) {
+        const textSection = generateTextSection(
+          markdown.substring(latestProcessedIndex, match.index)
```

Comment:
So this would imply that macros must always be top-level then, right? Because you're processing them as different markdown sections, you can't stop/resume in the middle of some construct (like a div or a table) to process a macro. Do you document anywhere how to/how not to use macros? (e.g. type supported, what you should/shouldn't do, etc.).